toolchain/lowering/lowering_block_worklist.h

```diff
@@ -0,0 +1,44 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_LOWERING_LOWERING_BLOCK_WORKLIST_H_
+#define CARBON_TOOLCHAIN_LOWERING_LOWERING_BLOCK_WORKLIST_H_
+
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/SmallVector.h"
+#include "toolchain/semantics/semantics_node.h"
+
+namespace Carbon {
+
+// A worklist for blocks that need to be lowered. Blocks are popped from the
+// worklist in the order in which they are first pushed, in a depth-first
+// traversal.
+class LoweringBlockWorklist {
+ public:
+  // Add a block to the work list, if it's not already present.
+  auto Push(SemanticsNodeBlockId id) -> void {
+    if (found_.insert(id).second) {
```

Comment:
In LoweringFunctionContext, you have `blocks_` which I think maps all found blocks. Maybe you could only `Push` entries not found in `blocks_`, removing `found_`?